File: ios/VCInvestmentDiscovery/Models/Company.swift
import Foundation

struct Company: Codable, Identifiable {
    let id: UUID
    let name: String
    let industry: String
    let foundingDate: Date
    let description: String
    let headquarters: String
    let website: String?
    var valuation: Double
    var employeeCount: Int
    let founders: [String]
    var stage: CompanyStage
    var fundingRounds: [FundingRound]
    var financialData: FinancialData
    var socialData: SocialData
    var marketData: MarketData

    init(id: UUID, name: String, industry: String, foundingDate: Date, description: String, headquarters: String, website: String?, valuation: Double, employeeCount: Int, founders: [String], stage: CompanyStage, fundingRounds: [FundingRound], financialData: FinancialData, socialData: SocialData, marketData: MarketData) {
        self.id = id
        self.name = name
        self.industry = industry
        self.foundingDate = foundingDate
        self.description = description
        self.headquarters = headquarters
        self.website = website
        self.valuation = valuation
        self.employeeCount = employeeCount
        self.founders = founders
        self.stage = stage
        self.fundingRounds = fundingRounds
        self.financialData = financialData
        self.socialData = socialData
        self.marketData = marketData
    }

    func totalFundingRaised() -> Double {
        return fundingRounds.reduce(0) { $0 + $1.amount }
    }

    func latestValuation() -> Double {
        return valuation
    }

    mutating func updateFinancialData(newData: FinancialData) {
        self.financialData = newData
    }

    mutating func updateSocialData(newData: SocialData) {
        self.socialData = newData
    }

    mutating func updateMarketData(newData: MarketData) {
        self.marketData = newData
    }
}

// Codable conformance is automatically synthesized by Swift
// Identifiable conformance is provided by the `id` property

// MARK: - Pending Human Tasks
// TODO: Implement data validation for company properties (e.g., non-negative valuation)
// TODO: Add methods for calculating growth rates and other key performance indicators
// TODO: Implement a mechanism to fetch and update company information from external APIs
// TODO: Create unit tests for the Company model and its methods
// TODO: Consider adding support for company logos and other media assets
// TODO: Implement a method to generate a company summary or elevator pitch

File: backend/src/models/Company.ts
import { Schema, model, Document } from 'mongoose';
import { CompanyStage, FinancialData, SocialData, MarketData } from '../types';

interface ICompany extends Document {
  _id: Schema.Types.ObjectId;
  name: string;
  industry: string;
  foundingDate: Date;
  description: string;
  headquarters: string;
  website: string;
  valuation: number;
  employeeCount: number;
  founders: string[];
  stage: CompanyStage;
  fundingRounds: Schema.Types.ObjectId[];
  financialData: FinancialData;
  socialData: SocialData;
  marketData: MarketData;

  totalFundingRaised(): Promise<number>;
  latestValuation(): number;
  updateFinancialData(newData: FinancialData): Promise<void>;
  updateSocialData(newData: SocialData): Promise<void>;
  updateMarketData(newData: MarketData): Promise<void>;
}

const CompanySchema = new Schema<ICompany>({
  name: { type: String, required: true },
  industry: { type: String, required: true },
  foundingDate: { type: Date, required: true },
  description: { type: String, required: true },
  headquarters: { type: String, required: true },
  website: { type: String },
  valuation: { type: Number, required: true },
  employeeCount: { type: Number, required: true },
  founders: { type: [String], required: true },
  stage: { type: String, enum: Object.values(CompanyStage), required: true },
  fundingRounds: { type: [Schema.Types.ObjectId], ref: 'FundingRound' },
  financialData: { type: Schema.Types.Mixed },
  socialData: { type: Schema.Types.Mixed },
  marketData: { type: Schema.Types.Mixed }
});

CompanySchema.methods.totalFundingRaised = async function(): Promise<number> {
  await this.populate('fundingRounds');
  return this.fundingRounds.reduce((total, round) => total + round.amount, 0);
};

CompanySchema.methods.latestValuation = function(): number {
  return this.valuation;
};

CompanySchema.methods.updateFinancialData = async function(newData: FinancialData): Promise<void> {
  this.financialData = newData;
  await this.save();
};

CompanySchema.methods.updateSocialData = async function(newData: SocialData): Promise<void> {
  this.socialData = newData;
  await this.save();
};

CompanySchema.methods.updateMarketData = async function(newData: MarketData): Promise<void> {
  this.marketData = newData;
  await this.save();
};

export default model<ICompany>('Company', CompanySchema);